File: tests/tainting_rules/go/library_access_map.go
package main

func handlerWritePos() {
	m := map[string]string{}
	m["body"] = source()
	m["user"] = "safe"
	// ruleid: taint
	sink(m["body"])
}

func handlerWriteNeg() {
	m := map[string]string{}
	m["body"] = "safe"
	m["user"] = source()
	// ok: taint
	sink(m["body"])
}

func handlerCommaOkPos() {
	m := map[string]string{}
	m["body"] = source()
	v, ok := m["body"]
	_ = ok
	// ruleid: taint
	sink(v)
}

func handlerCommaOkNeg() {
	m := map[string]string{}
	m["body"] = "safe"
	m["user"] = source()
	v, ok := m["body"]
	_ = ok
	// ok: taint
	sink(v)
}

func handlerDeleteClearsCell() {
	m := map[string]string{}
	m["body"] = source()
	delete(m, "body")
	// ok: taint
	sink(m["body"])
}

func handlerDeleteSiblingKeepsTaint() {
	m := map[string]string{}
	m["body"] = source()
	delete(m, "user")
	// ruleid: taint
	sink(m["body"])
}

File: tests/tainting_rules/scala/library_access_map.scala
import scala.collection.mutable

object T {
  def handlerPutPos(): Unit = {
    val m = mutable.Map[String, String]()
    m.put("body", source())
    m.put("user", "safe")
    // ruleid: taint
    sink(m.get("body"))
  }

  def handlerPutNeg(): Unit = {
    val m = mutable.Map[String, String]()
    m.put("body", "safe")
    m.put("user", source())
    // ok: taint
    sink(m.get("body"))
  }

  def handlerUpdatePos(): Unit = {
    val m = mutable.Map[String, String]()
    m.update("body", source())
    // ruleid: taint
    sink(m.get("body"))
  }

  def handlerApplyAssignPos(): Unit = {
    val m = mutable.Map[String, String]()
    m("body") = source()
    m("user") = "safe"
    // ruleid: taint
    sink(m.get("body"))
  }

  def handlerApplyAssignNeg(): Unit = {
    val m = mutable.Map[String, String]()
    m("body") = "safe"
    m("user") = source()
    // ok: taint
    sink(m.get("body"))
  }

  def handlerApplyReadPos(): Unit = {
    val m = mutable.Map[String, String]()
    m.put("body", source())
    // ruleid: taint
    sink(m("body"))
  }

  def handlerApplyReadNeg(): Unit = {
    val m = mutable.Map[String, String]()
    m.put("body", "safe")
    m.put("user", source())
    // ok: taint
    sink(m("body"))
  }

  def handlerGetOrElsePos(): Unit = {
    val m = mutable.Map[String, String]()
    m.put("body", source())
    // ruleid: taint
    sink(m.getOrElse("body", "fallback"))
  }

  def handlerGetOrElseTaintedDefault(): Unit = {
    val m = mutable.Map[String, String]()
    m.put("body", "safe")
    // ruleid: taint
    sink(m.getOrElse("body", source()))
  }

  def handlerPlusEqPos(): Unit = {
    val m = mutable.Map[String, String]()
    m += ("body" -> source())
    // ruleid: taint
    sink(m.get("body"))
  }

  def handlerPlusEqNeg(): Unit = {
    val m = mutable.Map[String, String]()
    m += ("body" -> "safe")
    m += ("user" -> source())
    // ok: taint
    sink(m.get("body"))
  }
}
